src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs

```diff
@@ -0,0 +1,289 @@
+                        current = ReadByte(ref span);
+                    }
+
+                    if (current == 'S')
+                    {
+                        ticks += value * TicksPerSecond;
+
+                        if (span.IsEmpty)
+                        {
+                            return Result(ticks);
+                        }
+
+                        current = ReadByte(ref span);
+                    }
+                }
+
+                if (current == '.' ||
+                    current == ',')
+                {
+                    ulong valueFraction = ReadDigits(ref span);
+                    (ulong ticksPerPart, ulong ticksPerFraction) = ReadByte(ref span) switch
+                    {
+                        (byte)'Y' => (TicksPerYear, TicksPerMonth),
+                        (byte)'M' => hasTime
+                            ? (TicksPerMinute, TicksPerSecond)
+                            : (TicksPerMonth, TicksPerDay),
+                        (byte)'D' => (TicksPerDay, TicksPerHour),
+                        (byte)'H' => (TicksPerHour, TicksPerMinute),
+                        (byte)'S' => (TicksPerSecond, 1U),
+                        _ => (0U, 0U),
+                    };
+
+                    if (span.IsEmpty && ticksPerPart != 0)
+                    {
+                        ulong scale = 10;
+                        while (valueFraction / scale != 0)
+                        {
+                            scale *= 10;
+                        }
+
+                        ulong ticksFraction = (ulong)(ticksPerPart * ((double)valueFraction / (double)scale));
+
+                        ticks += (ulong)(ticksFraction / ticksPerFraction) * ticksPerFraction;
+                        ticks += (ulong)(value * ticksPerPart);
+
+                        return Result(ticks);
+                    }
+                }
+            }
+
+            ThrowFormatException();
+            return default;
+
+            TimeSpan Result(ulong ticks) => new TimeSpan(negative ? -(long)ticks : (long)ticks);
+            static void ThrowFormatException() => throw ThrowHelper.GetFormatException(DataType.TimeSpan);
+        }
+
+        public override void Write(Utf8JsonWriter writer, TimeSpan value, JsonSerializerOptions options)
+        {
+            static int Write(Span<byte> span, char designator, ref ulong ticks, ulong ticksPerPart)
+            {
+                var value = ticks / ticksPerPart;
+                int written = 0;
+
+                if (value != 0)
+                {
+                    ticks -= value * ticksPerPart;
+                    written = WriteDigits(span, value);
+                    span[written++] = (byte)designator;
+
+                    return written;
+                }
+
+                return written;
+            }
+
+            static int WriteDigits(Span<byte> span, ulong value)
+            {
+                int digits = value switch
+                {
```

Comment:
the code for this seems currently less efficient: https://github.com/dotnet/runtime/issues/52730